### PR TITLE
feat: engine distributor for context

### PR DESCRIPTION
**Pull request by antmat**
This PR introduces several engines for distributing incoming (and outgoing) connections across threads. Current algo gives us improper distribution on near zero load (which is very bad for vicodyn or other service which do not perform reconnections).
This one introduces RR distributor and bucketed RR distributor as well as load distributor.

TODO:
  * [x] rename dispatcher to distributor
  * [x] extend tests
  * [x] intialize distributor in context via config
  * [x] remove duplicated code from distributor cpp

@3Hren @karitra PTAL

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -16,8 +16,10 @@ OPTION(COCAINE_ALLOW_TESTS "Build Tests" ON)
 OPTION(COCAINE_ALLOW_BENCHMARKS "Build Benchmarking Tools" OFF)
 
 # Import our CMake modules.
-INCLUDE(cmake/locate_library.cmake)
-INCLUDE(foreign/modules/AddCXXCompilerFlag.cmake)
+SET(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${PROJECT_SOURCE_DIR}/cmake)
+SET(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${PROJECT_SOURCE_DIR}/foreign/modules )
+INCLUDE(locate_library)
+INCLUDE(AddCXXCompilerFlag)
```
Comment (3Hren) [inline CMakeLists.txt:22]: The entire file is out of the scope. Split for separate PR's.

File: include/cocaine/format/exception.hpp [hunk 2]
```diff
@@ -17,7 +17,7 @@ struct display<std::exception> {
         if(ptr) {
             return stream << cocaine::format("[{}] {}", ptr->code().value(), ptr->what());
         } else {
-            return stream << ptr->what();
+            return stream << value.what();
```
Comment (3Hren) [inline include/cocaine/format/exception.hpp:20]: Out of the scope, but bugfixes (especially for templates) is ok.

File: src/context.cpp [hunk 3]
```diff
@@ -110,6 +111,7 @@ class context_impl_t : public context_t {
     // Service port mapping and pinning.
     port_mapping_t m_mapper;
 
+    std::unique_ptr<engine_dispatcher_t> engine_dispatcher;
```
Comment (3Hren) [inline src/context.cpp]: Where is it initialized? Also it's template as I see, where is its parameter?

File: src/context/engine_dispatcher.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
+#include "engine_dispatcher.hpp"
+
+namespace cocaine {
+
+struct load_engine_dispatcher_t: public engine_dispatcher_t {
+    auto
+    dispatch(const std::vector<std::unique_ptr<execution_unit_t>>& pool) -> execution_unit_t& {
+        typedef std::unique_ptr<execution_unit_t> unit_t;
```
Comment (3Hren) [inline src/context/engine_dispatcher.cpp]: Move to the superclass typedef, this covers the argument too.

File: src/context/engine_dispatcher.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    dispatch(const std::vector<std::unique_ptr<execution_unit_t>>& pool) -> execution_unit_t& {
+        static double max_value = 1.0;
+        static double bucket_size = 0.02;
+        size_t cur_bucket = max_value/bucket_size;
+        std::vector<std::reference_wrapper<execution_unit_t>> units;
+        units.reserve(pool.size());
+        for(auto& value: pool) {
+            size_t bucket = static_cast<size_t>(value->utilization()/bucket_size);
+            if(bucket < cur_bucket) {
+                cur_bucket = cur_bucket;
```
Comment (3Hren) [inline src/context/engine_dispatcher.cpp]: WTF?

File: src/context/engine_dispatcher.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                units.clear();
+                units.push_back(std::ref(*value));
+            } else if(bucket == cur_bucket) {
+                units.push_back(std::ref(*value));
+            }
+        }
```
Comment (3Hren) [inline src/context/engine_dispatcher.cpp]: No return?

File: src/context/engine_dispatcher.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        typedef std::unique_ptr<execution_unit_t> unit_t;
+        auto comp = [](const unit_t& lhs, const unit_t& rhs) {
+            return lhs->utilization() < rhs->utilization();
+        };
+        return **std::min_element(pool.begin(), pool.end(), comp);
+    }
+};
+
+
+struct rr_engine_dispatcher_t: public engine_dispatcher_t {
```
Comment (3Hren) [inline src/context/engine_dispatcher.cpp]: A few words how it works.

**Review by 3Hren — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (antmat) [reply CMakeLists.txt:22]: It's a bug fix - tests were unable to compile

Comment (3Hren) [reply CMakeLists.txt:22]: So you fixed tests, but disabled it?

File: src/context/distributor.hpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+#include "cocaine/errors.hpp"
+
+namespace cocaine {
+
+// distributor of elements of Pool
+// Pool should provide value_type typedef and 'operator[](size_t)' call
+// value_type should provide operator* and oprator->,
+// resulting type should provide 'double utilization()' call with returned values in [0.0 1.0] range
+template <class Pool>
+struct distributor {
```
Comment (3Hren) [inline src/context/distributor.hpp:14]: Missing virtual destructor.

File: src/context/distributor.hpp [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+};
+
+
+// Distributes elements of the Pool according only to 'utilization()' of the elements.
+// Simply returns the lowest value.
+template <class Pool>
+struct load_distributor: public distributor<Pool> {
+    using result_type = typename Pool::value_type;
+
+    load_distributor(const dynamic_t&) {}
```
Comment (3Hren) [inline src/context/distributor.hpp]: Pointless constructor.

File: src/context/distributor.hpp [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+};
+
+// Distributes elements of the Pool according to primitive round robin algorithm
+// Always returns element with next index (in a circular manner)
+template <class Pool>
+struct rr_distributor: public distributor<Pool> {
+    using result_type = typename Pool::value_type;
+    std::atomic_int counter;
+
+    rr_distributor(const dynamic_t&) : counter(0) {}
```
Comment (3Hren) [inline src/context/distributor.hpp]: Pointless argument.

File: src/context/distributor.hpp [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        return *std::min_element(pool.begin(), pool.end(), comp);
+    }
+};
+
+// Distributes elements of the Pool according to primitive round robin algorithm
+// Always returns element with next index (in a circular manner)
+template <class Pool>
+struct rr_distributor: public distributor<Pool> {
+    using result_type = typename Pool::value_type;
+    std::atomic_int counter;
```
Comment (3Hren) [inline src/context/distributor.hpp]: Public field that should be private.

File: src/context/distributor.hpp [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        return units[rand() % units.size()];
+    }
+};
+
+
+// Creates new distributor by name.
+template <class Pool>
+auto
+new_distributor(const std::string& name, const dynamic_t& args) -> std::unique_ptr<distributor<Pool>> {
```
Comment (3Hren) [inline src/context/distributor.hpp]: s/new/make

File: src/context/distributor.hpp [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    using ptr_t = std::unique_ptr<distributor<Pool>>;
```
Comment (3Hren) [inline src/context/distributor.hpp]: Use our fancy `make_unique`.

File: src/context/distributor.hpp [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    if(name == "load") {
+        return ptr_t(new load_distributor<Pool>(args));
+    } else if (name == "rr") {
+        return ptr_t(new rr_distributor<Pool>(args));
+    } else if (name == "bucket_random") {
+        return ptr_t(new bucket_random_distributor<Pool>(args));
+    }
+    throw error_t("unknown engine dispatcher type {}", name);
```
Comment (3Hren) [inline src/context/distributor.hpp]: Put in `` or "".

File: src/context/distributor.hpp [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    next(Pool& pool) -> result_type& override {
+        size_t cur_bucket = max_value/bucket_size;
+        std::vector<std::reference_wrapper<result_type>> units;
+        units.reserve(pool.size());
+        for(auto& value: pool) {
+            size_t bucket = static_cast<size_t>(value->utilization()/bucket_size);
+            if(bucket < cur_bucket) {
+                cur_bucket = bucket;
+                units.clear();
+                units.push_back(std::ref(value));
```
Comment (3Hren) [inline src/context/distributor.hpp]: Dup code.

File: src/context/distributor.hpp [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+            } else if(bucket == cur_bucket) {
+                units.push_back(std::ref(value));
```
Comment (3Hren) [inline src/context/distributor.hpp]: ^ with this.

File: src/context/distributor.hpp [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+    double bucket_size;
+
+    bucket_random_distributor(const dynamic_t& args) :
+        bucket_size(args.as_object().at("bucket_size", 0.02).as_double())
+    {}
+
+    auto
+    next(Pool& pool) -> result_type& override {
+        size_t cur_bucket = max_value/bucket_size;
+        std::vector<std::reference_wrapper<result_type>> units;
```
Comment (3Hren) [inline src/context/distributor.hpp]: I'd rather decompose this collection-bucket-of-minimal-load-engines into a separate method.

File: src/context/distributor.hpp [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+        units.reserve(pool.size());
+        for(auto& value: pool) {
+            size_t bucket = static_cast<size_t>(value->utilization()/bucket_size);
```
Comment (3Hren) [inline src/context/distributor.hpp]: auto

File: src/context/distributor.hpp [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+};
+
+// Split elements to several buckets according to utilization
+// Chooses bucket with lowest value and returns random element from it
+template <class Pool>
+struct bucket_random_distributor: public distributor<Pool> {
+    using result_type = typename Pool::value_type;
+
+    static constexpr const double max_value = 1.0;
+    double bucket_size;
```
Comment (3Hren) [inline src/context/distributor.hpp]: Public field that should be private.

**Review by 3Hren — CHANGES_REQUESTED** [review]

File: src/context.cpp [hunk 20]
```diff
@@ -187,6 +190,20 @@ class context_impl_t : public context_t {
         terminate();
     }
 
+    auto
+    initialize_distributor() -> void {
```
Comment (karitra) [inline src/context.cpp:194]: kind of aesthetics, but may be use free make_distributor(config_t&, logger_t&) function instead?

File: src/context/distributor.hpp [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+
+    auto
+    next(Pool& pool) -> result_type& override {
+        auto units = get_bucket_with_minimal_utilization(pool);
+        return units[rand() % units.size()];
+    }
+
+    auto
+    get_bucket_with_minimal_utilization(Pool& pool) -> bucket_t {
+        auto cur_bucket = max_value/bucket_size;
```
Comment (karitra) [inline src/context/distributor.hpp:101]: Is it ok for caller the case with `bucket_size == 0.0` ?

File: tests/unit/context.cpp [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+    auto operator* () const -> const engine_mock_t& {
+        return *this;
+    }
+
+    auto operator* () -> engine_mock_t& {
+        return *this;
+    }
+};
+
+TEST(context, bucket_random_dispatcher) {
```
Comment (karitra) [inline tests/unit/context.cpp:33]: Would be tests for marginal (or out of expected range) cases for init constants and utilization, e.g 0, 1,  1 < some_val or some_val < 0 useful?

File: src/context.cpp [hunk 23]
```diff
@@ -347,11 +364,7 @@ class context_impl_t : public context_t {
 
     execution_unit_t&
     engine() override {
-        typedef std::unique_ptr<execution_unit_t> unit_t;
-        auto comp = [](const unit_t& lhs, const unit_t& rhs) {
-            return lhs->utilization() < rhs->utilization();
-        };
-        return **std::min_element(m_pool.begin(), m_pool.end(), comp);
+        return *m_engine_distributor->next(m_pool);
```
Comment (karitra) [inline src/context.cpp:367]: What is expected count of execution_unit_t per pools? Would it be any profit of using heap based priority structures?

File: tests/unit/context.cpp [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+    auto d = make_distributor<std::vector<engine_mock_t>>("bucket_random", dynamic_t::object_t());
+    std::vector<double> values {
+        0.005, 0.002, 0.001, 0.15
+    };
+    std::vector<engine_mock_t> sample(values.begin(), values.end());
+    std::map<double, size_t> counter;
```
Comment (karitra) [inline tests/unit/context.cpp:39]: Use of double as a key always looks horrid, but may be ok here.

File: src/context/distributor.hpp [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+// Always returns element with next index (in a circular manner)
+template <class Pool>
+struct rr_distributor: public distributor<Pool> {
+    using result_type = typename Pool::value_type;
+
+    rr_distributor() : counter(0) {}
+
+    auto
+    next(Pool& pool) -> result_type& override {
+        return pool[counter++ % pool.size()];
```
Comment (karitra) [inline src/context/distributor.hpp]: Would empty pool behave neat here?

File: src/context/distributor.hpp [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+    static constexpr const double max_value = 1.0;
+
+    bucket_random_distributor(const dynamic_t& args) :
+        bucket_size(args.as_object().at("bucket_size", 0.02).as_double())
+    {}
+
+    auto
+    next(Pool& pool) -> result_type& override {
+        auto units = get_bucket_with_minimal_utilization(pool);
+        return units[rand() % units.size()];
```
Comment (karitra) [inline src/context/distributor.hpp]: Is it ok to obtain empty `units` sequence here?

Comment (karitra) [reply src/context/distributor.hpp]: Is it `std::rand` here? What guaranties are on maximum random value?

File: src/context/distributor.hpp [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+    using result_type = typename Pool::value_type;
+
+    load_distributor() {}
+
+    auto
+    next(Pool& pool) -> result_type& override {
+        auto comp = [](const result_type& lhs, const result_type& rhs) {
+            return lhs->utilization() < rhs->utilization();
+        };
+        return *std::min_element(pool.begin(), pool.end(), comp);
```
Comment (karitra) [inline src/context/distributor.hpp]: If pool is (unexpectedly) empty it would be reference to pool::end here.

**Review by karitra — APPROVED** [review]
LGTM and awesome, in general, except `bucket_random_distributor` machinery would need attentive rereading a bit later.

File: src/context.cpp (hunk 23 above)
Comment (karitra) [reply src/context.cpp:367]: It seems that it could be possibility for m_pool to be empty on incorrect `m_config->network().pool()` parameter, but may be (hopefully) it is some kind of assertion somewhere.

File: src/context.cpp (hunk 20 above)
Comment (antmat) [reply src/context.cpp:194]: There is config check in this function (for backward compatibility)

File: src/context/distributor.hpp (hunk 27 above)
Comment (antmat) [reply src/context/distributor.hpp]: Pool can not be empty afaik

File: src/context.cpp (hunk 23 above)
Comment (antmat) [reply src/context.cpp:367]:
```
            m_pool     = source.at("pool", boost::thread::hardware_concurrency() * 2).as_uint();

            if(m_pool <= 0) {
                throw cocaine::error_t("network I/O pool size must be positive");
            }
```

File: src/context/distributor.hpp (hunk 25 above)
Comment (antmat) [reply src/context/distributor.hpp]: Same - pool can not be empty

File: src/context/distributor.hpp (hunk 26 above)
Comment (antmat) [reply src/context/distributor.hpp]: units should not be empty as far as pool is not empty. 
yep, it's std::rand, I think it's good enough for us returning values in range of [0, max int32]

File: src/context/distributor.hpp (hunk 21 above)
Comment (antmat) [reply src/context/distributor.hpp:101]: good catch, I'll add check

**Comment by 3Hren** [comment]
Merged in b00dba20cf853bcb9736c3bc05ed9cd6395680de.
